### ISSUE TITLE
Where is the CognosNT.exe and COGNOSAPI.dll source code?

### ISSUE THREAD

**Issue by nitz** [thradde/FastWinDirStat#4]
Both of CognosNT.exe and COGNOSAPI.dll are distributed through the [installer located here](https://www.idealsoftware.com/opensource/Fast%20WinDirStat%20Setup.exe), and is referenced [in the source that is available](https://github.com/thradde/FastWinDirStat/blob/95a58c408a2202509f83840f47884ef61c3a7876/make/windirstat/CogFileFind.cpp#L60). However, the source for those binaries is missing. As FastWinDirStat is distributed under the GPL v2 as the original was, the source for all modules must be distributed as well.

**Comment by thradde** [comment]
This is wrong. CognosNT.exe is a separate executable, which is not under GPL. GPL refers to static or dynamic linking, which is not the case.

**Comment by nitz** [comment]
Please excuse my mistake on the CognosNT.exe front. It seems that I'm missing something though, but the only license I see is the GPL v2:

![image](https://github.com/thradde/FastWinDirStat/assets/602691/27d69702-0e61-4c2f-8fe7-e574c8d7ad88)

If CognosNT.exe is under some other license, it's not stated anywhere in the package that I can find.

However with COGNOSAPI.dll being dynamically linked, it should still be covered by the GPL.

**Comment by thradde** [comment]
The closed-source license for CognosNT is stated in the readme.md. IDEAL Software GmbH grants an exclusive license for use in FastWinDirStat. It is also wrong that libraries linked into GPL'ed software need to be open source. The GPLs intention is that you don't derive closed source from open source, but not vice versa. Take for example kernel32.dll. It is closed source, but every open source program is using it. I hope I made things clear now.

**Comment by nitz** [comment]
I believe you're mistaken. Kernel32.dll is covered with a specific clause in the license:

> However, as a special exception, the source code distributed need not include anything that is normally distributed (in either source or binary form) with the major components (compiler, kernel, and so on) of the operating system on which the executable runs, unless that component itself accompanies the executable.

As a component that is included as part of the operating system on which the executable runs, and that it does not accompany the executable itself, it is except from needing its source to be distributed. COGNOSAPI.dll does not meet those requirements and thus would not qualify for the exemption.

**Comment by thradde** [comment]
Please quote the part of the GPL v2, where it is written that calling an added external DLL must be open source.

**Comment by nitz** [comment]
via [GPL v2.0](https://www.gnu.org/licenses/old-licenses/gpl-2.0.en.html), 


Section 0:

> [...] The "Program", below, refers to any such program or work, and a "work based on the Program" means either the Program or any derivative work under copyright law: that is to say, a work containing the Program or a portion of it, either verbatim or with modifications and/or translated into another language. (Hereinafter, translation is included without limitation in the term "modification".) [...]

Section 2: 

> You may modify your copy or copies of the Program or any portion of it, thus forming a work based on the Program, [...]

> [...] But when you distribute the same sections as part of a whole which is a work based on the Program, the distribution of the whole must be on the terms of this License, whose permissions for other licensees extend to the entire whole, and thus to each and every part regardless of who wrote it. 

Section 3:

> You may copy and distribute the Program (or a work based on it, under Section 2) in object code or executable form under the terms of Sections 1 and 2 above provided that you also do one of the following: 
>
>  a) Accompany it with the complete corresponding machine-readable source code, which must be distributed under the terms of Sections 1 and 2 above on a medium customarily used for software interchange; or, 
>
>  b) Accompany it with a written offer, valid for at least three years, to give any third party, for a charge no more than your cost of physically performing source distribution, a complete machine-readable copy of the corresponding source code, to be distributed under the terms of Sections 1 and 2 above on a medium customarily used for software interchange; or, 
>
>  c) Accompany it with the information you received as to the offer to distribute corresponding source code. (This alternative is allowed only for noncommercial distribution and only if you received the program in object code or executable form with such an offer, in accord with Subsection b above.) 
>
> The source code for a work means the preferred form of the work for making modifications to it. For an executable work, complete source code means all the source code for all modules it contains, plus any associated interface definition files, plus the scripts used to control compilation and installation of the executable. [...]

**Comment by thradde** [comment]
Thank you for your effort. You know, when you deal with legal questions, you always have to ask, what the intention of a contractual regulation is. The intention of the GPL - as I already wrote - is, not to derive work from open source and make it closed source, so that someone benefits commerially from the work of others. This is not the case with FastWinDirStat.

I would also like to point out this paragraph of the GPL v2:
"These requirements apply to the modified work as a whole. If identifiable sections of that work are not derived from the Program, and can be reasonably considered independent and separate works in themselves, then this License, and its terms, do not apply to those sections when you distribute them as separate works. But when you distribute the same sections as part of a whole which is a work based on the Program, the distribution of the whole must be on the terms of this License, whose permissions for other licensees extend to the entire whole, and thus to each and every part regardless of who wrote it."

So, the solution would be to make two installers, one for FastWinDirStat and one for CognosNT. As a result they are not distributed as a whole. By the way, FastWinDirStat continues to work without CognosNT (you can disable the CognosNT service in the Windows service manager). In this case it will fall back to the original code.

But you know, I am too lazy to create two installers. I rather will wait what the FSF (or to whom you reported this issue) will say about it. In the worst case I will delete this repository, so the bugfixes, the 64-bit version, the new installer and the (little) enhanced functionality will be lost for the public - if this makes you happy.

**Comment by rfl890** [comment]
Is CognosNT your proprietary file scanner? Why not open-source it, is it used in any commercial projects?
